File: receptor_crc-32.py
import zlib
import socket

# CAPA 5: TRANSMISIÓN
def recibir_trama(host='localhost', puerto=5000):
    print("--- CAPA DE TRANSMISIÓN ---")
    servidor = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    servidor.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    servidor.bind((host, puerto))
    servidor.listen(1)
    print(f"Escuchando en {host}:{puerto}...")
    cliente, direccion = servidor.accept()
    print(f"Conexión establecida desde {direccion}")
    datos = b''
    while True:
        parte = cliente.recv(4096)
        if not parte:
            break
        datos += parte
    cliente.close()
    print("Trama recibida.")
    try:
        trama = datos.decode()
    except UnicodeDecodeError:
        print("Error: No se pudo decodificar la trama recibida.")
        return None
    return trama

# CAPA 3: ENLACE
def verificar_integridad(trama_recibida):
    print("\n--- CAPA DE ENLACE ---")
    if trama_recibida is None:
        return None, None, True
    if len(trama_recibida) < 32:
        print("Error: La trama recibida es demasiado corta para contener CRC.")
        return None, None, True
    mensaje_binario = trama_recibida[:-32]
    crc_recibido = trama_recibida[-32:]
    # Reconstruir texto para cálculo de CRC
    mensaje_texto = ""
    for i in range(0, len(mensaje_binario), 8):
        byte = mensaje_binario[i:i+8]
        mensaje_texto += chr(int(byte, 2))
    valor_crc_calc = zlib.crc32(mensaje_texto.encode())
    crc_calc_bin = format(valor_crc_calc & 0xFFFFFFFF, '032b')
    print(f"CRC recibido: {crc_recibido}")
    print(f"CRC calculado: {crc_calc_bin}")
    error = crc_calc_bin != crc_recibido
    if error:
        print("Error detectado en la transmisión (CRC diferente).")
    else:
        print("No se detectaron errores (CRC coincide).")
    return mensaje_binario, mensaje_texto, error

# CAPA 2: PRESENTACIÓN y CAPA 1: APLICACIÓN
def procesar_mensaje(mensaje_binario, mensaje_texto, error):
    print("\n--- CAPA DE PRESENTACIÓN ---")
    if error:
        print("No es posible decodificar el mensaje debido a errores detectados.")
        return
    # Decodificar binario a mensaje
    mensaje_decodificado = ""
    for i in range(0, len(mensaje_binario), 8):
        byte = mensaje_binario[i:i+8]
        mensaje_decodificado += chr(int(byte, 2))
    print("--- CAPA DE APLICACIÓN ---")
    print("Mensaje recibido correctamente:")
    print(mensaje_decodificado)

if __name__ == "__main__":
    print("--- RECEPTOR INICIADO ---")
    trama = recibir_trama()
    mensaje_binario, mensaje_texto, error = verificar_integridad(trama)
    procesar_mensaje(mensaje_binario, mensaje_texto, error)

File: ReceptorCRC32.java
// ReceptorCRC32.java
import java.io.*;
import java.net.*;
import java.util.zip.CRC32;

public class ReceptorCRC32 {

    // CAPA 5: TRANSMISIÓN
    public static String recibirTrama(String host, int puerto) throws IOException {
        System.out.println("--- CAPA DE TRANSMISIÓN ---");
        ServerSocket server = new ServerSocket();
        server.setReuseAddress(true);
        server.bind(new InetSocketAddress(host, puerto));
        System.out.printf("Escuchando en %s:%d...%n", host, puerto);

        Socket client = server.accept();
        System.out.printf("Conexión establecida desde %s%n", client.getRemoteSocketAddress());

        BufferedReader in = new BufferedReader(new InputStreamReader(client.getInputStream()));
        StringBuilder sb = new StringBuilder();
        String line;
        // asume que toda la trama viene en una sola línea
        while ((line = in.readLine()) != null) {
            sb.append(line);
        }
        client.close();
        server.close();

        System.out.println("Trama recibida.");
        return sb.toString();
    }

    // CAPA 3: ENLACE
    public static Result verificarIntegridad(String trama) {
        System.out.println("\n--- CAPA DE ENLACE ---");
        if (trama == null || trama.length() < 32) {
            System.err.println("Error: trama inválida o demasiado corta.");
            return new Result(null, null, true);
        }

        String mensajeBin = trama.substring(0, trama.length() - 32);
        String crcRecibido = trama.substring(trama.length() - 32);
        System.out.println("CRC recibido:  " + crcRecibido);

        // reconvertir binario a texto
        StringBuilder texto = new StringBuilder();
        for (int i = 0; i < mensajeBin.length(); i += 8) {
            String byteStr = mensajeBin.substring(i, i + 8);
            texto.append((char) Integer.parseInt(byteStr, 2));
        }

        // calcular CRC32
        CRC32 crc32 = new CRC32();
        crc32.update(texto.toString().getBytes());
        String crcCalc = String.format("%32s", Long.toBinaryString(crc32.getValue() & 0xFFFFFFFFL))
                             .replace(' ', '0');
        System.out.println("CRC calculado: " + crcCalc);

        boolean error = !crcCalc.equals(crcRecibido);
        System.out.println(error
            ? "Error detectado en la transmisión (CRC diferente)."
            : "No se detectaron errores (CRC coincide).");

        return new Result(mensajeBin, texto.toString(), error);
    }

    // CAPA 2+1: PRESENTACIÓN y APLICACIÓN
    public static void procesarMensaje(Result r) {
        System.out.println("\n--- CAPA DE PRESENTACIÓN ---");
        if (r.error) {
            System.out.println("No es posible decodificar el mensaje debido a errores detectados.");
            return;
        }
        System.out.println("--- CAPA DE APLICACIÓN ---");
        System.out.println("Mensaje recibido correctamente:");
        System.out.println(r.texto);
    }

    public static void main(String[] args) throws IOException {
        System.out.println("--- RECEPTOR INICIADO ---");
        String trama = recibirTrama("localhost", 5000);
        Result r = verificarIntegridad(trama);
        procesarMensaje(r);
    }

    // Helper para retornar múltiples valores
    private static class Result {
        String binario, texto;
        boolean error;
        Result(String b, String t, boolean e) {
            binario = b; texto = t; error = e;
        }
    }
}
